cpp/src/Ice/Incoming.cpp

```diff
@@ -657,45 +617,65 @@ IceInternal::Incoming::invoke(const ServantManagerPtr& servantManager, InputStre
         }
     }
 
+    const bool amd = false;
     try
     {
-        //
-        // Dispatch in the incoming call
-        //
-        _servant->_iceDispatch(*this, _current);
-
-        //
-        // If the request was not dispatched asynchronously, send the response.
-        //
-        if(!_inAsync)
+        // Dispatch the request to the servant.
+        if (_servant->_iceDispatch(*this))
         {
-            response(false);
+            // If the request was dispatched synchronously, send the response.
+            response(amd);
         }
     }
     catch (...)
     {
-        if(_inAsync)
-        {
-            try
-            {
-                _inAsync->kill(*this);
-            }
-            catch (const Ice::ResponseSentException&)
-            {
-                const Ice::PropertiesPtr properties = _os.instance()->initializationData().properties;
-                if(properties->getPropertyAsIntWithDefault("Ice.Warn.Dispatch", 1) > 0)
-                {
-                    warning(current_exception());
-                }
-                return;
-            }
-        }
-        exception(current_exception(), false);
+        // An async dispatch is not allowed to throw any exception because it moves "this" memory into a new Incoming
+        // object.
+        exception(current_exception(), amd);
     }
 }
 
-const Current&
-IceInternal::IncomingRequest::getCurrent()
+void
+Incoming::completed()
 {
-    return _in.getCurrent();
+    const bool amd = true;
+    setResponseSent();
+    response(amd);
+}
+
+void
+Incoming::completed(exception_ptr ex)
+{
+    const bool amd = true;
+    setResponseSent();
+    exception(ex, amd);
+}
+
+void
+Incoming::failed(exception_ptr ex) noexcept
+{
+    try
+    {
+        completed(ex);
+    }
+    catch (...)
+    {
+        // Ignore all exceptions. The caller in the dispatch thread can't handle any exception because its memory
+        // was moved to a new "async" incoming object.
+    }
+}
+
+void
+Incoming::setResponseSent()
```

Comment:
See above, I'm pretty sure this can be removed.

Reply:
See above.